src/dmd/backend/cgxmm.d

```diff
@@ -1573,6 +1576,11 @@ bool xmmIsAligned(elem *e)
            )
             return false;       // definitely not aligned
     }
+    else if (e.Eoper == OPind)
+    {
+        // can't know for sure until runtime
+        return false;   // assume not aligned
```

Comment:
First we can't really assume random pointers to be aligned to any boundary.

But to illustrate why we should be pessimistic here consider the following program:
```D
import core.simd;
void main()
{
    struct S { long4 y; }
    S v; v.y += 1; // crashes
}
```
if you compile this little program with the current DMD and run it, it will crash obscurely, usually it will say something like `Illegal instruction`. If you run the program on [Intel's emulator](https://software.intel.com/en-us/articles/intel-software-development-emulator/) you'll get a more useful error message. For example on my machine it says:
```
SDE ERROR:  TID: 0 executed instruction with an unaligned memory reference to address 0x7ffffddfc8b0 INSTR: 0x55ce337f463a: IFORM: VMOVDQA_MEMqq_YMMqq :: vmovdqa ymmword ptr [rbp-0x20], ymm1
	IMAGE:    /home/user/simd
	FUNCTION: _Dmain
	FUNCTION ADDR: 0x55ce337f460c
```
It complains about unaligned memory. However the address `rbp-32` which is is this case `0x7ffffddfc8b0` is in fact aligned to `16` bytes (`0x7ffffddfc8b0 % 16 == 0`) because it is required for the stack to be aligned to `16` bytes upon function entry (on Linux 64) but not aligned required to be aligned to `32` bytes thus our stack pointer isn't aligned to `32` (`0x7ffffddfc8b0 % 32 == 16`).

The CPU does not seem to tolerate misaligned pointers with instructions which expect aligned pointer therefor it is safer to use the slower instructions that work with both aligned and misaligned pointers.

Reply:
As for stack variable, perhaps we can force a 32 bytes stack alignment, this is what LDC does. LDC generates an instruction in the prolog of the function which looks like this:
```asm
and    rsp,0xffffffffffffffe0     # 0xffffffffffffffe0 == ~(32 - 1)
```

Reply:
I moved this into a separate pull request #9143